File: csharp/Day3.cs
using System.Text.RegularExpressions;

namespace AdventOfCode2024;

public class Day3
{
    public static void Run()
    {
        var countPart1 = 0;
        var countPart2 = 0;
        var input = File.ReadAllText("../../../../csharp/day3.txt").Replace("\n", "");
        var matches = Regex.Matches(input, @"(do[(][)]|don't[(][)]|mul[(]([0-9]+),([0-9]+)[)])");
        var mech = "do()";
        foreach (Match match in matches)
        {
            if (match.Groups[1].Value.StartsWith('d'))
            {
                mech = match.Groups[1].Value;
                continue;
            }
            var number = int.Parse(match.Groups[2].Value) * int.Parse(match.Groups[3].Value);
            countPart1 += number;
            if (mech == "do()") countPart2 += number;
        }
        Console.WriteLine($"Part 1: {countPart1}");
        Console.WriteLine($"Part 2: {countPart2}");
    }
}

File: python/day3.py
import re
def calculate(nums):
    return nums[0] * nums[1]
for string in {open('day3.txt', 'r').read(), re.sub(r"((don't[(][)].*?)+(do[(][)]))|(don't(.*?)$)", "" , open('day3.txt', 'r').read().replace("\n", ""))}: 
    print(sum([calculate(list(map(int, re.findall(r"[0-9]+", nums)))) for nums in re.findall(r"mul[(][0-9]+,[0-9]+[)]", string)]))